### ISSUE TITLE
Plans for Dropwizard 3

### ISSUE THREAD

**Issue by zUniQueX** [dropwizard/dropwizard#4608]
Including Logback's alpha versions, the dependencies could now be updated to the `jakarta` namespace.

Should we - as most other Jakarta EE components - support both the `javax` and the `jakarta` namespace and therefore build Dropwizard 3, or should we postpone the upgrade?

**Comment by zUniQueX** [comment]
At the moment, the upgrade seems to get really heavy, since Jetty is on version 9 yet. Unfortunately we cannot upgrade Jetty to 10.0.x due to a logback issue. A PR for this is already raised: https://github.com/qos-ch/logback/pull/532

**Comment by zUniQueX** [comment]
@joschi For an easier upgrade to the `jakarta` namespace I took a look at an upgrade of Jetty to 10.x.x. [This commit](https://github.com/zUniQueX/dropwizard/commit/b5e713b27f7b490ab61c7e78f022896d55321cba) does the job, however I'm not quite pleased with it. Logback hereby is the current blocking factor, since the `JettyServerAdapter` does not support `HttpFields.Mutable`. I introduced a temporary custom `Jetty10ServerAdapter`, but therefore also had to implement the `AppenderAttachableImplExposingRequestLogImpl`, which makes me unhappy about this.
Since I don't know, when the referenced PR will get merged, I see this as an alternative anyway, if we would want to do the big bang change.

Should I provide a PR for this? The newly introduced classes are marked as deprecated and will be removed when using the Logback classes that support the `jakarta` namespace. I just think, that without the Jetty upgrade the major upgrade of Dropwizard to v3 will get quite messy.

**Comment by pstackle** [comment]
I wonder if we should be considering an approach similar to what Jetty did with their v10 and v11 split. We would do Dropwizard v3 synced up with Jetty 10 and stay on the `javax` namespace and simultaneously create v4 that syncs up with Jetty 11 and uses the `jakarta` namespace. 

Ref: https://webtide.com/jetty-10-and-11-have-arrived/

CC @dropwizard/committers

**Comment by zUniQueX** [comment]
> I wonder if we should be considering an approach similar to what Jetty did with their v10 and v11 split. We would do Dropwizard v3 synced up with Jetty 10 and stay on the `javax` namespace and simultaneously create v4 that syncs up with Jetty 11 and uses the `jakarta` namespace. 
> 
> Ref: https://webtide.com/jetty-10-and-11-have-arrived/
> 
> CC @dropwizard/committers 

@pstackle I think this would be nice for users but will be very hard to maintain. Keeping only two versions in sync with the `javax` and `jakarta` namespaces gets a difficult task. And with this configuration we would have to port changes also for the three different Jetty versions.

**Comment by pstackle** [comment]
@zUniQueX What do you mean by _three_ Jetty versions?

I was thinking we'd have:
- Dropwizard v3 (targeted at Jetty 10 supporting the `javax` namespace)
- Dropwizard v4 (targeted at Jetty 11 supporting the `jakarta` namespace)

with Dropwizard v2 (and Jetty 9) no longer supported or at least in a (hopefully) low-effort maintenance mode.

---
Alternatively, if we don't mirror the approach Jetty has done, we'd basically have to pick between the following choices:
- Dropwizard v3 targets Jetty 10, continues using `javax` namespace and users have to wait longer to complete the `javax` to `jakarta` migration until some time in the future when a new Dropwizard version targets Jetty 11+
- Dropwizard v3 targets Jetty 11 (skips Jetty 10), users have a hard cutover from (Jetty 9 + `javax`) to (Jetty 11 + `jakarta`)

**Comment by zUniQueX** [comment]
@pstackle I exactly meant Jetty 9 in v2, Jetty 10 in v3 and Jetty 11 in v4. I think, we cannot directly deprecate v2.

In my opinion we can skip a Jetty 10 only version, so I would stay on Jetty 9 in v2 and upgrade to Jetty 11 in v3.

**Comment by joschi** [comment]
@dropwizard/committers @zUniQueX What do you think about the following roadmap?

- Dropwizard 2.2.x with Jetty 10.x (and Java 11 baseline)
- Dropwizard 3.x with Jetty 11.x (and Java 11 baseline and Jakarta EE namespace)

While Dropwizard 2.2.x formally is breaking semantic versioning (because of the change of the language baseline from Java 8 to 11), it would still work largely as before and it might give users a little nudge to upgrade to Java 11 in production. :wink:

Other than that I don't think Jetty 10 will break a lot of things which are officially supported by Dropwizard (see also https://webtide.com/jetty-10-and-11-have-arrived/) and thus the upgrade from Dropwizard 2.1.x to 2.2.x should cause only minor upgrade pain - if at all when people are already on Java 11.

**Comment by joschi** [comment]
@danjee No, this is currently only in planning stage and there's no development branch yet.

**Comment by danjee** [comment]
@joschi thanks. I've edited the question

**Comment by joschi** [comment]
@danjee Moving from `javax.*` to `jakarta.*` is a _huge_ effort and we need to upgrade basically all libraries which are using `javax.*` right now.

This is not only Jersey and Jetty, but also Hibernate Validator, Hibernate ORM, and any library adding functionality to those aforementioned ones.

**Comment by mattnelson** [comment]
> * Dropwizard 2.2.x with Jetty 10.x (and Java 11 baseline)
> * Dropwizard 3.x with Jetty 11.x (and Java 11 baseline and Jakarta EE namespace)
> 
> While Dropwizard 2.2.x formally is breaking semantic versioning (because of the change of the language baseline from Java 8 to 11), it would still work largely as before and it might give users a little nudge to upgrade to Java 11 in production.

Would request documented timelines on end of support for 2.x, 2.1.x, and 2.2.x so consumers can start planning around it as soon as possible.

**Comment by zUniQueX** [comment]
> @dropwizard/committers @zUniQueX What do you think about the following roadmap?
> 
> * Dropwizard 2.2.x with Jetty 10.x (and Java 11 baseline)
> * Dropwizard 3.x with Jetty 11.x (and Java 11 baseline and Jakarta EE namespace)
>
>While Dropwizard 2.2.x formally is breaking semantic versioning (because of the change of the language baseline from Java 8 to 11), it would still work largely as before and it might give users a little nudge to upgrade to Java 11 in production. 😉

@joschi From other PRs and issues I thought you would highly prefer semantic versioning and therefore wouldn't want to make  such an upgrade in 2.2.x :wink: I personally agree Java 11 upgrades should be done for production environments and therefore would appreciate the change.

For the upgrade effort I previously linked a [test branch](https://github.com/zUniQueX/dropwizard/commit/b5e713b27f7b490ab61c7e78f022896d55321cba) I have for this.
Currently I see these points to note for an upgrade:
 - `addEventListener`: The upgrade to the `addEventListener` method should be not much a problem, as the compiler would complain about this.
 - GZIP User-Agent exclusion: As exclusion of User-Agents got removed, the property in the configuration is just unnecessary
 - GZIP default accepted methods: from Jetty 10 onwards, `GET` and `POST` are enabled by default. This also shouldn't be a suprising change to users, but could be nice to note.
 - Additional bean in `LifecycleEnvironment`: `LifeCycleEnvironment` now doesn't only contain the `JettyManaged` objects.
 - SNI host checking: Jetty 10 uses are more strict implementation of server-side SNI host checking in the `SecureRequestsCustomizer`. Therefore previously valid client configurations will now fail. But I didn't want to set the `disableSniHostCheck` property in the default configuration.
 - `shutdownGracePeriod` in HTTP/2 tests: I couldn't find out why the HTTP/2 integration tests would only fail sometimes in various executions. In the logs can be seen, that the request was already done, when an exception occurred. Disabling graceful shutdown seems to fix the problem, so I see this as a reasonable workaround for now. If any user does a similar thing, this could be very suprising.

Currently the limiting factor of doing an upgrade seems to be that Logback still doesn't support Jetty 10. We could work around this with a custom implementation of the `JettyServerAdapter`, but I don't know, if we _should_ do this.

**Comment by zUniQueX** [comment]
> Hello, is there a dropwizard 3 preview repository available? I am only interested in the jakarta libs upgrade. I've downloaded and tried to build the project on the renovate-jakarta branch but got some errors there, I've updated the jersey libs but still got into some problems with jackson which uses old javax packages. They had a bom with jakarta classifierd but revert it. Looks like it's a long way to have it.

IIRC the only necessary Jackson dependency for this upgrade has a Jakarta version with the `jackson-jakarta-json-provider`.

**Comment by zUniQueX** [comment]
> @danjee Moving from `javax.*` to `jakarta.*` is a _huge_ effort and we need to upgrade basically all libraries which are using `javax.*` right now.
> 
> This is not only Jersey and Jetty, but also Hibernate Validator, Hibernate ORM, and any library adding functionality to those aforementioned ones.
> 
> In other words: Don't hold your breath. 😉

@joschi Would you be interested in having an early development branch for DW 3? After the Jetty 10 upgrade this is more or less replacing libraries and doing search-and-replace in the IDE :wink: I've done this before and could complete the `mvn install`. For more I just didn't find the time. I could upgrade this branch to the current master and provide a PR, if there would be interest in this.

**Comment by joschi** [comment]
> From other PRs and issues I thought you would highly prefer semantic versioning and therefore wouldn't want to make such an upgrade in 2.2.x 😉

@zUniQueX And you'd be absolutely right about it. 😅

Thanks to yours and @pstackle's feedback I'm convinced that the jump to Jetty 10 and Java 11 makes more sense for Dropwizard 3.x and the migration to Jakarta EE in Dropwizard 4.x. 👍
